Fb coupling time

ukmo-juan-castillo opened the pull request:
Branch allowing a coupling time step different than the model time step. If no coupling time step is declared in the namelist, it will be made equal to the model time step after throwing a warning. The coupling time step must be a multiple of the model time step, otherwise the program will throw and error and stop.

A single, new regtest has been added in ww3_tp2.14 to test the new functionality. This has been tested in the ukmo_cray computer, for which a few changes had to be made in the regtests scripts to allow running these coupling test. A minor bug fix in the ice coupling regtests has also been fixed. The rest of the regtests also passed without problems.

The only thing missing would be adding a comment in the code informing of the changes, for which I would have to know the version for this branch (supposedly 7.08)

ukmo-juan-castillo committed:
First set of changes - code is ready but regtests still to be completed

ukmo-juan-castillo committed:
Regtests for non-default oasis time step added, tested in ukmo_cray;
minor bug fixes to coupled regtests

ukmo-juan-castillo commented:
I only left Chris and Andy as reviewers, more should be added once they are happy with the changes, see ticket 229:

https://github.com/NOAA-EMC/WW3/issues/229

ukmo-juan-castillo commented:
One more modification that is possibly needed is updating the manual, explicitly saying what the default and valid values of the coupling time step are.

ukmo-ccbunney commented on regtests/bin/matrix_ukmo_cray:
```diff
@@ -141,6 +142,13 @@ fi
 # 2.  Execute matrix.base ...                                                 #
 # --------------------------------------------------------------------------- #
              
+# Compiler configuration, required for coupling
```
@ukmo-juan-castillo 
Can you move these `cmplr` lines into the `$comp` conditional sections starting at line 63, please?
You can then delete line 87 that is actually set incorrectly (not by you).

ukmo-juan-castillo replied:
This change has been now addressed and has been pushed in the fb branch. Waiting for comments from other reviewers regarding the changes in matrix.base.

ukmo-ccbunney reviewed (changes requested):
Just a small change request from me and a comment:
I see that you have included the coupling regtest (ww3_tp2.14) in the `matrix.base`.
I note that this was not part of the matrix.base before and I wonder if this was for a reason? @aliabdolali, @mickaelaccensi - are you able to comment on this?

ukmo-ansaulter reviewed (approved):
Thanks Juan,

Subject to the minor change suggested by @ukmo-ccbunney I'm happy with these changes.

Looking at the manual, the entry for OASIS it is a little thin, and I also don't think that an example of the namcouple namelist exists in the repository. It would be good to add these, but can I suggest as part of a new issue to review the OASIS manual entry more widely? Issue to be associated with the NOAA milestone: https://github.com/NOAA-EMC/WW3/milestone/22 

As we are thinking of this change as a bugfix, the version of ww3_shel targeted should be the latest one

ukmo-juan-castillo committed:
The Met Office reviewers' comments have been addressed

ukmo-juan-castillo commented:
For running the new regtests added in ww3_tp2.14, new input files would need to be added to the NOAA ftp site:

r-ww3.nc.OASACM2
r-toy.nc.OASACM2
toy_coupled_field.nc.OASACM2

It is sufficient if these files are a copy of the equivalent files for OASACM, although only 11 of the times contained in the files will be used.

JessicaMeixner-NOAA commented:
> For running the new regtests added in ww3_tp2.14, new input files would need to be added to the NOAA ftp site:
> 
> r-ww3.nc.OASACM2
> r-toy.nc.OASACM2
> toy_coupled_field.nc.OASACM2
> 
> It is sufficient if these files are a copy of the equivalent files for OASACM, although only 11 of the times contained in the files will be used.

If can use the existing files already on the ftp, you can just copy them to more than one location in the script here: https://github.com/NOAA-EMC/WW3/blob/develop/model/bin/ww3_from_ftp.sh#L44-L45 instead of adding duplications of the file on the ftp server.

ukmo-juan-castillo committed:
Updated the script that copies the data necessary for running the
new regtests

mickaelaccensi commented:
could you remove the option -c $cmplr from matrix.base since it's already defined in $rtst ?
In matrix_datarmor and matrix_ncep , cmplr is not an environment variable, we don't have an export cmplr.

   echo "$rtst -s OASACM -w work_OASACM -c $cmplr -C OASIS -f -p $mpi -n $np -o netcdf $ww3 ww3_tp2.14" >> matrix.body
   echo "$rtst -s OASACM2 -w work_OASACM2 -c $cmplr -C OASIS -f -p $mpi -n $np -o netcdf $ww3 ww3_tp2.14" >> matrix.body
   echo "$rtst -s OASOCM -w work_OASOCM -c $cmplr -C OASIS -f -p $mpi -n $np -o netcdf $ww3 ww3_tp2.14" >> matrix.body
   echo "$rtst -s OASICM -w work_OASICM -c $cmplr -C OASIS -f -p $mpi -n $np -o netcdf $ww3 ww3_tp2.14" >> matrix.body

ukmo-juan-castillo committed:
Changes to ukmo_cray regtests as suggested by reviewer

ukmo-juan-castillo commented:
The latest suggested change has been implemented and tested. The trick to make everything work afterwards was setting by default the variable cmplOption to 'y' in matrix_ukmo_cray.

mickaelaccensi commented:
Would you mind to remove this line from ww3_tp2.14/input/toy/prep_env.sh
`sed -i "s/icc/gcc/g" cmplr` 
and add a new sed statement on wwatch3_cc like that :
`     sed -e "s/<optc_short>/$optcs/" -e "s/<optl_short>/$optls/" -e "s/<comp_mpi>/$comp_mpi/" -e "/<wwatch3_cc>/$WWATCH3_CC/" -e "s/<comp_mpi_exe>/$comp_mpi_exe/" cmplr.tmpl > cmplr`
and add this at line18 (still in ww3_tp2.14/input/toy/prep_env.sh) (take care to add the quotes, it's interpreted here)
`export WWATCH3_CC=grep WWATCH3_CC $WWATCH3_ENV | awk -F' ' '{print $2}'`

then, change in ww3_tp2.14/input/oasis3-mct/util/make_dir/cmplr.tmpl :
`CC          = icc`
by
`CC          = <wwatch3_cc>`

ukmo-juan-castillo committed:
Updates to the coupling regtests as suggested by reviewer

ukmo-juan-castillo commented:
Latest changes suggested by the reviewer, which improve the coupling regtests scripts, have been implemented and tested.

ukmo-ccbunney commented:
@mickaelaccensi - Are you happy that the ww3_tp2.14 regtest is now included as part of the matrix.base? It was not included before - I just wanted to check if there was a reason for that or not (it is a more complex test case for certain!)

ukmo-ccbunney commented:
@ukmo-juan-castillo Can you please target the `staging` branch for this PR (not develop)? Thanks.

ukmo-juan-castillo committed:
Merge branch 'develop' into fb_coupling_time

ukmo-juan-castillo committed:
Merge branch 'staging' into fb_coupling_time

mickaelaccensi commented:
> @mickaelaccensi - Are you happy that the ww3_tp2.14 regtest is now included as part of the matrix.base? It was not included before - I just wanted to check if there was a reason for that or not (it is a more complex test case for certain!)

Yes @ukmo-ccbunney  ! I'm really happy with that and even more to know that you are also using OASIS for model coupling so we can work together to improve it.

mickaelaccensi commented:
@ukmo-juan-castillo , I think there is a typo in matrix.base for calendar (the end of the line is missing): 
`  echo "  echo '      Calendar type                      : $calendar'"         >> matrix.head`

and in matrix_datarmor, matrix_ncep, matrix_zeus_HLT, the export calendar is missing : 
`    export calendar='y' # Calendar type`

ukmo-juan-castillo committed:
Update matrix.base

Fixed after error pointed out by reviewer

mickaelaccensi commented:
can you also update the matrix_* files ? 

in matrix_datarmor, matrix_ncep, matrix_zeus_HLT, the export calendar is missing :
`export calendar='y' # Calendar type`

ukmo-juan-castillo committed:
Update matrix_datarmor

ukmo-juan-castillo committed:
Update matrix_ncep

ukmo-juan-castillo committed:
Update matrix_zeus_HLT

mickaelaccensi commented:
do you also have differences in OUTPUT_TOY.txt between this branch and develop branch ?
>After decomp_def, il_paral =            1        4400         400
< After decomp_def, il_paral =            1        4800         400

and also differences in the values sent by the toy model, I don't know why.

in both tests, the APPLE partitionning is used but it provide different partitions. There is no changes in oasis3_mct directory neither in toymodel directory.

Do you see an explanation about it ?

ukmo-juan-castillo commented:
I will investigate these differences, thanks for pointing them out.

ukmo-juan-castillo commented:
The problem is that the output of the toy_model program is not bit reproducible. This happens because it is a MPI program, but the piece of code that writes to output does not take into account this fact.

I have two different solutions: one that writes just one output file, and other that writes one output file for each processor. I imagine you would like to implement here one of these fixes, which one would you prefer?

ukmo-juan-castillo committed:
Fix non-reproducible toy output in coupled regtests

ukmo-juan-castillo committed:
Merge branch 'fb_coupling_time' of https://github.com/ukmo-waves/WW3 into fb_coupling_time

ukmo-juan-castillo commented:
I modified the toy program so that it only produces an output file for processor zero, as I believe this will be sufficient to check there are no differences between runs. I modified the matrix.comp script accordingly to stop comparing oasis logs and to remove time and location differences in the output.ww3 file (log file of WWIII when running in coupled mode).

mickaelaccensi commented:
ww3_tp2.14 are bit4bit now. thanks a lot Juan for this work. I approve the PR

ukmo-ccbunney commented:
@JessicaMeixner-NOAA : are you happy with these changes? If so, I will get this PR signed off and merged. Thanks!

JessicaMeixner-NOAA commented on regtests/bin/matrix_ncep:
```diff
@@ -154,6 +154,8 @@ fi
   export   infgrv='y' # Second harmonic generation tests
   export     uost='y' # ww3_ts4 Unresolved Obstacles Source Term (UOST)
   export    assim='y' # Restart spectra update
+  export     coup='y' # Atmosphere, ocean, and ice coupling
```
Are there any secondary library requirements for running these tests?  Also we should specify that this is OASIS coupling, not just "coupling".

JessicaMeixner-NOAA replied:
I'm asking about the requirements, because if there are any, I think this should be a 'n' for NCEP.

JessicaMeixner-NOAA commented on regtests/bin/matrix.base:
```diff
@@ -85,6 +85,7 @@
   echo "  echo '      Multi 07 (unstr. + reg. grds)      : $multi07'"          >> matrix.head
   echo "  echo '      Multi 08 (with ice)                : $multi08'"          >> matrix.head
   echo "  echo '      Assim Update of the restart file   : $assim'"            >> matrix.head
+  echo "  echo '      Atmosphere, ocean, and ice coupling: $coup'"             >> matrix.head
```
I'd prefer that we specify that these are OASIS coupling as there are multiple ways to couple and this is only for OASIS testing.

JessicaMeixner-NOAA commented on regtests/bin/matrix.base:
```diff
@@ -2003,6 +2004,16 @@
   fi
 
+ #Test of atmosphere, ocean, and ice coupling 
+ if [ "$coup" = 'y' ]
```
Since these are all MPI cases, we should make sure dist=y as well.

JessicaMeixner-NOAA commented on regtests/bin/matrix.base:
```diff
@@ -2003,6 +2004,16 @@
+ 
 # 365_day and 360_day calendars, no switch sharing here
 
   if [ "$calendar" = 'y' ] && [ "$shrd" = 'y' ]
```
This does not need the "$shrd"=y as there is not an alternative dist option.

ukmo-juan-castillo replied:
I have modified the new regtests to specify that they are for OASIS coupling, and they will only run when dist='y'. All the dependencies are internal, i.e., all the libraries needed are already contained and compiled in the regtests without further external libraries. For that reason, I think it is a good idea to leave them in NCEP.

For the moment I leave out the modifications for calendar, as they belong to another ticket, and because imposing shrd='y' is only redundant. I can easily change it if you insist on modifying this here, though.

JessicaMeixner-NOAA reviewed (commented):
I have never used the COU switch or any OASIS coupling, so I do not have any comments on that code or updates.   If the OASIS coupling tests require extra libraries I think NCEP's matrix should turn these off.     My only comment is that in the regression testing that its specified that these are OASIS coupling tests and that the dist=y be added in matrix.base to protect against those wanting to only run serial tests.

ukmo-juan-castillo committed:
Specify that the new coupled regtests are for OASIS coupling, and run
them only wi mpi mode

ukmo-ccbunney commented on regtests/ww3_tp2.14/input/prep_env.sh:
```diff
@@ -36,7 +37,8 @@ then
      [ "$cmplr" == "zeus_pgi" ] || [ "$cmplr" == "zeus_pgi_debug" ]               || \
-     [ "$cmplr" == "datarmor_pgi" ] || [ "$cmplr" == "datarmor_pgi_debug" ] ; then
+     [ "$cmplr" == "datarmor_pgi" ] || [ "$cmplr" == "datarmor_pgi_debug" ]       || \
+     [ "$cmplr" == "ukmo_cray_gnu" ] || [ "$cmplr" == "ukmo_cray_gnu_debug" ] ; then
```
@ukmo-juan-castillo Could you add an entry for `ukmo_cray` compiler here to keep it consistent with the `cmplr.env` file? I.e:
```
[ "$cmplr" == "ukmo_cray" ] || [ "$cmplr" == "ukmo_cray_debug" ]   || \\
```

ukmo-juan-castillo replied:
The issue with the calendar was in matrix.base:

  if [ "$calendar" = 'y' ] && [ "$shrd" = 'y' ]
  then

should be:

  if [ "$calendar" = 'y' ]
  then

because the second condition is redundant. I will update it now that I have to commit again, and also update the prep_env.sh script.

ukmo-ccbunney replied:
Thanks!

ukmo-ccbunney reviewed (changes requested):
@ukmo-juan-castillo  - Just a couple of things.
 1. A small addition to `prep_env.sh` script.
 2. I see that there is till a `calendar = "y"` entry in the regtest matrix files. We're you planning on keeping that? You comment above suggested you were removing it (as it relates to a different PR).

ukmo-juan-castillo committed:
Minor changes to add the ukmo_cray compiler to the coupling test, and to
remove reduntant condition in matrix.base for calendar tests

ukmo-ccbunney commented:
All merged - thanks @ukmo-juan-castillo.
Many thanks also to @JessicaMeixner-NOAA and @mickaelaccensi for their thorough reviews. 👍
